From 0685103bd2600cfd4467b1f5ec181777090b866b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Kacper=20Michaj=C5=82ow?= <kasper93@gmail.com>
Date: Tue, 29 Sep 2026 11:34:13 +0200
Subject: [PATCH] avformat/hls: free the previous segments if parsing a
 playlist fails
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Kacper Michajłow <kasper93@gmail.com>
(cherry picked from commit 0685103bd2600cfd4467b1f5ec181777090b866b)

Hades: the two frees move after fail: as upstream; in 9.0.2
last_load_time is still set before fail:, which b5402d1ecf90 (not taken)
changed upstream. The change is upstream's.
---
 libavformat/hls.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/libavformat/hls.c b/libavformat/hls.c
index afef11b77a14..e0d61e033fa9 100644
--- a/libavformat/hls.c
+++ b/libavformat/hls.c
@@ -1141,13 +1141,13 @@ static int parse_playlist(HLSContext *c, const char *url,
             av_log(c->ctx, AV_LOG_WARNING, "Media sequence changed unexpectedly: %"PRId64" -> %"PRId64"\n",
                    prev_start_seq_no, pls->start_seq_no);
         }
-        free_segment_dynarray(prev_segments, prev_n_segments);
-        av_freep(&prev_segments);
     }
     if (pls)
         pls->last_load_time = av_gettime_relative();
 
 fail:
+    free_segment_dynarray(prev_segments, prev_n_segments);
+    av_freep(&prev_segments);
     av_free(new_url);
     if (close_in)
         ff_format_io_close(c->ctx, &in);
